File: src/embedded/traceBoundary.cpp
#include "anki/embeddedVision.h"

namespace Anki
{
  namespace Embedded
  {
#define CHECK_FOR_OUT_OF_BOUNDS
#define NUM_DIRECTIONS 8

    const Point_s16 offsets[NUM_DIRECTIONS] = {Point_s16(0,-1), Point_s16(1,-1), Point_s16(1,0), Point_s16(1,1), Point_s16(0,1), Point_s16(-1,1), Point_s16(-1,0), Point_s16(-1,-1)};
    //const s32 dxs[NUM_DIRECTIONS] = {0, 1, 1, 1, 0, -1, -1, -1};
    //const s32 dys[NUM_DIRECTIONS] = {-1, -1, 0, 1, 1, 1, 0, -1};

    //function newDirection = findNewDirection(img, curPoint, curDirection, value)
    static BoundaryDirection FindNewDirection(const Array_u8 &binaryImg, const Point_s16 &curPoint, const BoundaryDirection curDirection, const u8 value)
    {
      BoundaryDirection newDirection = BOUNDARY_UNKNOWN;

      s32 directionList1Limits[2];
      s32 directionList2Limits[2];
      if(curDirection == 0) {
        directionList1Limits[0] = NUM_DIRECTIONS-1;
        directionList1Limits[1] = NUM_DIRECTIONS;
        directionList2Limits[0] = 0;
        directionList2Limits[1] = NUM_DIRECTIONS-2;
      } else {
        directionList1Limits[0] = curDirection-1;
        directionList1Limits[1] = NUM_DIRECTIONS;
        directionList2Limits[0] = 0;
        directionList2Limits[1] = curDirection-1;
      }

      for(s32 newDirection=directionList1Limits[0]; newDirection<directionList1Limits[1]; newDirection++) {
        //const u8 imgValue = *(binaryImg.Pointer(curPoint.y+dys[newDirection], curPoint.x+dxs[newDirection]));
        const u8 imgValue = *(binaryImg.Pointer(curPoint + offsets[newDirection]));
        if(imgValue == value) {
          return static_cast<BoundaryDirection>(newDirection);
        }
      }

      for(s32 newDirection=directionList2Limits[0]; newDirection<directionList2Limits[1]; newDirection++) {
        //const u8 imgValue = *(binaryImg.Pointer(curPoint.y+dys[newDirection], curPoint.x+dxs[newDirection]));
        const u8 imgValue = *(binaryImg.Pointer(curPoint + offsets[newDirection]));
        if(imgValue == value) {
          return static_cast<BoundaryDirection>(newDirection);
        }
      }

      return BOUNDARY_UNKNOWN;
    }

    Result TraceBoundary(const Array_u8 &binaryImg, const Point_s16 &startPoint, BoundaryDirection initialDirection, FixedLengthList_Point_s16 &boundary)
    {
#if ANKI_DEBUG_LEVEL == ANKI_DEBUG_HIGH
      DASConditionalErrorAndReturnValue(binaryImg.IsValid(),
        RESULT_FAIL, "TraceBoundary", "binaryImg is not valid");

      DASConditionalErrorAndReturnValue(boundary.IsValid(),
        RESULT_FAIL, "TraceBoundary", "boundary is not valid");

      // Is the start point inside the inner pixel of the image?
      DASConditionalErrorAndReturnValue(
        startPoint.x > 0 &&
        startPoint.y > 0 &&
        static_cast<s32>(startPoint.x) < (binaryImg.get_size(1)-1) &&
        static_cast<s32>(startPoint.y) < (binaryImg.get_size(0)-1),
        RESULT_FAIL, "TraceBoundary", "startPoint is out of bounds");

      DASConditionalErrorAndReturnValue(boundary.get_maximumSize() == MAX_BOUNDARY_LENGTH,
        RESULT_FAIL, "TraceBoundary", "boundary must be have a maximum size of MAX_BOUNDARY_LENGTH");
#endif // #if ANKI_DEBUG_LEVEL == ANKI_DEBUG_HIGH

      boundary.Clear();

      if(initialDirection == BOUNDARY_UNKNOWN)
        initialDirection = BOUNDARY_N;

      //if ~checkForOutOfBounds
      //    % The boundaries are zero, which means we don't have to check for out of bounds
      //    binaryImg([1,end],:) = 0;
      //    binaryImg(:, [1,end]) = 0;
      //end

      //curDirection = initialDirection;
      BoundaryDirection curDirection = initialDirection;

      //curPoint = [startPoint(1), startPoint(2)];
      Point_s16 curPoint(startPoint);

#if ANKI_DEBUG_LEVEL == ANKI_DEBUG_HIGH
      DASConditionalErrorAndReturnValue(*(binaryImg.Pointer(curPoint)) != 0,
        RESULT_FAIL, "TraceBoundary", "startPoint must be on a non-zero pixel of binaryImg");
#endif // #if ANKI_DEBUG_LEVEL == ANKI_DEBUG_HIGH

      // printf("0) %d %d\n", curPoint.y+1, curPoint.x+1);

      boundary.PushBack(startPoint);

      //while true
      for(s32 i=1; i<MAX_BOUNDARY_LENGTH; i++) {
        //% Search counter-clockwise until we find a 0
        const BoundaryDirection zeroDirection = FindNewDirection(binaryImg, curPoint, curDirection, 0);

        if(zeroDirection == BOUNDARY_UNKNOWN)
          break;

        //% Search counter-clockwise until we find a 1
        const s32 zeroDirectionPlusOneS32 = static_cast<s32>(zeroDirection) + 1;
        const BoundaryDirection zeroDirectionPlusOne = (zeroDirectionPlusOneS32==NUM_DIRECTIONS) ? static_cast<BoundaryDirection>(0) : static_cast<BoundaryDirection>(zeroDirectionPlusOneS32);

        const BoundaryDirection oneDirection = FindNewDirection(binaryImg, curPoint, zeroDirectionPlusOne, 1);

        //newPoint = [curPoint(1)+dy(oneDirection+1), curPoint(2)+dx(oneDirection+1)];
        //const Point2<u16> newPoint(curPoint.x+dxs[oneDirection], curPoint.y+dys[oneDirection]);
        const Point_s16 newPoint(curPoint + offsets[oneDirection]);

        // printf("%d) %d %d\n", i, newPoint.y+1, newPoint.x+1);

#ifdef CHECK_FOR_OUT_OF_BOUNDS
        /*if(newPoint.x < 0 &&
        newPoint.y < 0 &&
        newPoint.x >= binaryImg.get_size(1) &&
        newPoint.y >= binaryImg.get_size(0)) {
        return RESULT_OK;
        }*/
        if(newPoint.x <= 0 &&
          newPoint.y <= 0 &&
          static_cast<s32>(newPoint.x) >= (binaryImg.get_size(1)-1) &&
          static_cast<s32>(newPoint.y) >= (binaryImg.get_size(0)-1)) {
            return RESULT_OK;
        }
#endif //#ifdef CHECK_FOR_OUT_OF_BOUNDS

        //% If the new point is the same as the last point (in other words, if the search is stuck on a point), we're done
        if(newPoint == *boundary.Pointer(i-1)) {
          return RESULT_OK;
        }

        boundary.PushBack(newPoint);

        //% If the new point is the same as the first point, quit.
        if(newPoint == *boundary.Pointer(0)) {
          return RESULT_OK;
        }

        if(i > 1 &&
          (abs(newPoint.x - boundary.Pointer(0)->x) + abs(newPoint.y - boundary.Pointer(0)->y)) <= 1) {
            return RESULT_OK;
        }

        curPoint = newPoint;
      } // for(s32 i=0; i<MAX_BOUNDARY_LENGTH; i++) {
      return RESULT_OK;
    }
  } // namespace Embedded
} // namespace Anki

File: coretech/vision/matlab/traceBoundary.m

% function [boundary, boundaryImg] = traceBoundary(binaryImg, startPoint, initialDirection)

% May work incorrectly on an object that is less than two pixels thick
% If the start point is in a corner, the boundary may circle the object twice

% img = ones(10,10); img(4:6, 4:6) = 0;
% [boundary, boundaryImg] = traceBoundary(img, [3,7], 'n')

% img2 = ones(16,16); img2(6:8, 6:8) = 0; img2(4, 4:10) = 0; img2(3:10, 10) = 0; img2(10, 4:10) = 0; img2(3:10, 4) = 0;
% [boundary2, boundaryImg2] = traceBoundary(img2, [7,9], 'n')

% img3 = ones(16,16); img3(6:8, 6:8) = 0; img3(3, 4:10) = 0; img3(3:10, 11) = 0; img3(11, 4:10) = 0; img3(3:10, 3) = 0;
% [boundary3, boundaryImg3] = traceBoundary(img3, [7,9], 'n')

% img4 = ones(16,16); img4(6:8, 6:8) = 0; img4(3, 4:10) = 0; img4(3:10, 11) = 0; img4(11, 4:10) = 0; img4(3:10, 3) = 0; img4(4:5, 7) = 0;
% [boundary4, boundaryImg4] = traceBoundary(img4, [7,9], 'n')

% img5 = ones(16,16); img5(6:8, 6:8) = 0; img5(3, 4:11) = 0; img5(3:10, 11) = 0; img5(11, 3:11) = 0; img5(3:10, 3) = 0; img5(4:5, 7) = 0;
% [boundary5, boundaryImg5] = traceBoundary(img5, [7,9], 'n')

function [boundary, boundaryImg] = traceBoundary(binaryImg, startPoint, initialDirection)

validDirectionStrings = {'n','ne','e','se','s','sw','w','nw'};
initialDirection = find(strcmpi(initialDirection, validDirectionStrings) == 1, 1) - 1;

if isempty(initialDirection)
    return;
end

if nargout == 2
    boundaryImg = zeros(size(binaryImg));
end

checkForOutOfBounds = false;

if ~checkForOutOfBounds
    % The boundaries are zero, which means we don't have to check for out of bounds
    binaryImg([1,end],:) = 0;
    binaryImg(:, [1,end]) = 0;
end

allDirections = 0:7;
dx = [0, 1, 1, 1, 0, -1, -1, -1];
dy = [-1, -1, 0, 1, 1, 1, 0, -1];

curDirection = initialDirection;

boundary = [startPoint(1), startPoint(2)];
curPoint = [startPoint(1), startPoint(2)];

if nargout == 2
    curBoundaryIndex = 1;
    boundaryImg(curPoint(1), curPoint(2)) = curBoundaryIndex;
end

while size(boundary,1) < 5000
    % Search counter-clockwise until we find a 0
   zeroDirection = findNewDirection(binaryImg, curPoint, curDirection, 0);
   if zeroDirection == -1
       break;
   end
   
    % Search counter-clockwise until we find a 1
    zeroDirectionPlusOne = zeroDirection + 1;
    if zeroDirectionPlusOne == 8
        zeroDirectionPlusOne = 0;
    end
    
    oneDirection = findNewDirection(binaryImg, curPoint, zeroDirectionPlusOne, 1);
        
    newPoint = [curPoint(1)+dy(oneDirection+1), curPoint(2)+dx(oneDirection+1)];
        
    curPoint = newPoint;
    boundary(end+1, :) = newPoint;
    if nargout == 2
        curBoundaryIndex = curBoundaryIndex + 1;
        boundaryImg(newPoint(1), newPoint(2)) = curBoundaryIndex;
    end
    
    % If the new point is the same as the last point (or second-to-last to detect oscillations), quit.
    if sum(newPoint == boundary(end-1,:)) == 2 ...
        || (size(boundary,1)>2 && sum(newPoint == boundary(end-2,:)) == 2)
        break;
    end
    
    % If the new point is the same as the first point, quit.
%     keyboard
    if sum(newPoint == boundary(1,:)) == 2 || ...
        (size(boundary,1)>2 && sum(abs(newPoint-boundary(1,:))) <= 1)
%         keyboard
        break;
    end
    
    if checkForOutOfBounds
        % If the new point is out of bounds, quit.
        if newPoint(1)<=1 || newPoint(2)<=1 ||...
           newPoint(1)>=size(binaryImg,1) || newPoint(2)>=size(binaryImg,2)
            break;
        end
    end
end % while true

if size(boundary,1) > 4000
    keyboard;
end

end % function boundary = traceBoundary(binaryImg, startPoint, initialDirection)

function newDirection = findNewDirection(img, curPoint, curDirection, value)
    allDirections = 0:7;
    dx = [0, 1, 1, 1, 0, -1, -1, -1];
    dy = [-1, -1, 0, 1, 1, 1, 0, -1];

    newDirection = -1;
    
    if curDirection == 0
        searchDirections = [7, 0:6];
    else
        searchDirections = [(curDirection-1):(length(allDirections)-1), 0:(curDirection-2)];
    end
    
    for direction = searchDirections
        if img(curPoint(1)+dy(direction+1), curPoint(2)+dx(direction+1)) == value
            newDirection = direction;
            break;
        end
    end
    
%     keyboard
end

